### PR TITLE
Mixins

### PR DESCRIPTION
**Pull request by edmundnoble**
Still doesn't work, the types aren't being fully substituted. In particular, the parameters of the copied methods have the wrong types.

### CODE REVIEW

File: plugin/src/main/scala/scalaz/plugin/Definitions.scala [hunk 1]
```diff
@@ -20,9 +20,11 @@ abstract class Definitions {
   lazy val ScalazMetaPackage = ensurePackage(ScalazPackage.moduleClass, "meta")
 
   // TODO
+  lazy val MixinAttr: ClassSymbol      = rootMirror.getRequiredClass("scalaz.meta.mixin")
```
Comment (hrhino) [inline plugin/src/main/scala/scalaz/plugin/Definitions.scala]: setting us up the bomb?

File: plugin/src/main/scala/scalaz/plugin/Mixins.scala [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+import scala.tools.nsc.typechecker.AnalyzerPlugins
+import scala.tools.nsc.{Global, Phase, plugins}
+import scala.util.control.NonFatal
+
+abstract class Mixins
+    extends plugins.PluginComponent {
+  val global: Global
+  val scalazDefns: Definitions {val global: Mixins.this.global.type}
+
+  implicit final class Ops[A](val self: A) {
```
Comment (hrhino) [inline plugin/src/main/scala/scalaz/plugin/Mixins.scala]: this could be in `Utils`

File: plugin/src/main/scala/scalaz/plugin/Mixins.scala [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    def opt[B](f: PartialFunction[A, B]): Option[B] = f.lift(self)
+  }
+
+  val mixinPlugin = new global.analyzer.MacroPlugin {
+    override def isActive(): Boolean =
+      global.phase.id < global.currentRun.picklerPhase.id
+
+    val state: mutable.ListMap[String, StatePart] = mutable.ListMap.empty
+
+    type TyParamSubstMap = Map[String, global.Type]
```
Comment (hrhino) [inline plugin/src/main/scala/scalaz/plugin/Mixins.scala]: do you ever use the type? I think these could be one `ParamSubstMap = Map[Name, Symbol]` (where `Name` disambiguates type/value params)

**Comment by edmundnoble** [comment]
Addressed review, but I still need to fix conflicts, add more tests, and better error handling.

**Pull request merged**